torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,255 @@
-  Variable() : Tensor() {}
-  Variable(const Variable & rhs) : Tensor(rhs) {}
-  Variable(Variable && rhs) noexcept : Tensor(std::move(rhs)) {}
+  /// Attempt to get a pointer to the gradient accumulator of the `Variable`,
+  /// if it still exists. If the gradient accumulator function has been
+  /// destroyed, returns a `nullptr`.
+  std::shared_ptr<Function> try_get_grad_accumulator() const;
 
-  // Implicitly casts a Tensor to a Variable. This should only be called on
-  // Tensors which you know are actually Variables.
-  /*implicit*/ Variable(Tensor const & rhs) : Tensor(rhs) {}
-  /*implicit*/ Variable(Tensor && rhs) noexcept : Tensor(std::move(rhs)) {}
+  /// Get the gradient accumulator of the `Variable` if it has one, or else
+  /// create one on the fly and return it.
+  std::shared_ptr<Function> grad_accumulator() const;
 
+  /// Set the gradient edge -- i.e. `grad_fn` and `input_nr` -- of the
+  /// `Variable`.
+  /// NOTE: This will always set the `grad_fn`, even if this is a leaf
+  /// variable, and never the `grad_accumulator`. For the latter, use
+  /// `set_grad_accumulator`. This allows late construction of an interior
+  /// `Variable`.
+  void set_gradient_edge(Edge&& edge);
+
+  /// Return the "canonical" gradient edge of this `Variable`, i.e. either the
+  /// gradient function if this is an interior `Variable`, or the gradient
+  /// accumulator otherwise. If the `Variable` is interior, the returned `Edge`
+  /// will store the input index of the `Function` to which this variable is
+  /// connected in its `input_nr` field. For leaves, the `input_nr` is always
+  /// zero. Note that `set_gradient_edge` and `gradient_edge` are not
+  /// symmetric. You must use `set_gradient_edge` to set the `grad_fn` and
+  /// `set_grad_accumulator` to set the accumulator.
   Edge gradient_edge() const {
     // If grad_fn is null (as is the case for a leaf node), we instead
-    // interpret the gradient function to be a grad accumulator,
-    // which will accumulate its inputs into the grad property of the
-    // variable. These nodes get suppressed in some situations,
-    // see "suppress grad accumulation" below. Note that only variables which
-    // have `requires_grad = True` can have grad accumulators.
+    // interpret the gradient function to be a gradient accumulator, which will
+    // accumulate its inputs into the grad property of the variable. These
+    // nodes get suppressed in some situations, see "suppress gradient
+    // accumulation" below. Note that only variables which have `requires_grad =
+    // True` can have gradient accumulators.
     if (const auto& gradient = grad_fn()) {
       return Edge(gradient, output_nr());
     } else {
       return Edge(grad_accumulator(), 0);
     }
   }
 
-  inline VariableImpl* get() const;
+  /// Return the input index of the gradient `Function` to which this `Variable`
+  /// is connected.
+  int output_nr() const noexcept;
```

Comment:
The advantage of marking these declarations inline in the header is that you get compilation errors (instead of runtime or linker errors) if you miss the definition below.

Reply:
Ah interesting, didn't know that. I felt like it was a bit of "visual noise" in the declarations. Do you think we should put them back?